text/0000-debuginfo-based-panic.md

```diff
@@ -0,0 +1,115 @@
+
+Furthermore, there is no automated solution right now that could remove panic location strings from binaries. This is a problem for developers of closed-source software as it can, in some circumstances, faciliate reverse engineering.
+
+Because including debug info in release binaries is not viable for various reasons (most notably debug info is often *larger* than code), many platforms already store it in a separate location (`*.pdb` on Windows, `*.dSYM` on MacOS). On Linux and several other unixoid platforms, this approach is still quite new and uncommon but nonetheless perfectly possible. 
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+* The Rust compiler simply continues to generate platform-compatible debug information. It gains a new switch that causes it to generate *external* debug info. What this means is that the binary itself is stripped - all debug info is written to a separate **symbol file** (in a platform-specific format).
+* Cargo's release profile enables external debug info.
+* Panic messages now always include a backtrace (instead of the old location string).
+	* This means that the root cause of panics from `unwrap()` and similar interfaces are now much clearer.
+	* It's simple and obvious as most modern languages print backtraces for unhandled errors.
+	* The backtrace is generated from debug information. Without debug symbols, memory addresses are printed instead. The standard `addr2line` utility can be used to manually obtain location information once debug info is available.
+	* User perspectives:
+		* You're developing an application. You build in the `dev` profile. You have debug symbols and thus complete backtraces for every panic.
+		* You're testing your application (think QA). It's a `release` build. You have debug symbols and thus complete backtraces for every panic.
+		* You ship your application to an end user. You ship only the binary, not the symbol file. The end user runs into a panic and receives a raw stack trace (of memory addresses). The end user now reports the bug to you. Using `addr2line` you obtain a complete backtrace.
+		* Alternatively, you ship the symbol file to the user. The user can now debug your application without recompiling and - obviously - gets complete backtraces for every panic.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+**Symbol files** already exist (and work) on Windows and MacOS. On Linux, it's as simple as writing the main binary to one ELF image and all the debug sections to another. The existing `-C debuginfo` compiler options gains a third level "3 = full debug info in an external symbol file". The cargo release profile sets `debuginfo=3` (instead of `debuginfo=0`).
```

Comment:
The location of the debuginfo is orthogonal to its detail. Why couldn't I have out-of-binary level 1 debuginfo? It seems like it would make more sense for this to be a separate flag.

Reply:
Rationale here: https://github.com/rust-lang/rfcs/pull/2154#discussion_r140089310

I'm not very happy with this either, I don't care about it. The only reason this isn't a new compiler flag is that I was too lazy to write up the whole "new flag, -Z/-C, stabilization, Cargo.toml" boilerplate.